### ISSUE TITLE
"Bookmark" connection settings for quick access

### ISSUE THREAD

**Issue by bennigraf** [rabbitControl/rcp-ts-client#15]
Hey there!

As mentioned in my PR #14 recently, I thought it would be nice to be able to store connection settings somehow, similar to bookmarks in browsers. That way one wouldn't have to remember connection settings everytime they want to use some rabbit-controlleable entity via this web app, even if they had multiple such entities under their control.

I did some experimentation directly in code and came up with a solution that would work for me :-). I'll add a screen capture to the end of this post for reference. I want to put this up for discussion first – let me know what you think! If you think this is fine I could file another PR here.

Some notes:
 - I use `localStorage` as storage backend, as that's generally available on all modern devices (as far as I know).
 - Note how the bookmark list will expand only when there are three or more bookmarks saved.
 - I refrained from adding some logic checking for "duplicate" bookmarks after some thinking because it seemed to create more complications than it would solve. So currently it's perfectly possible to create multiple identical bookmarks with the exact same settings; it's up to the user to deal with that. I don't see any problems with that, though.
 - I did not fully understand what "Tabs in Root" means; however, I can easily store that info in the bookmarks as well, obviously
 - I added the "disconnect" button mainly for my experiment ;-)... I'd leave it in if you don't mind, even though in real world usage it might not be necessary to ever actually disconnect...

Best,
Benjamin.

![bookmarking](https://user-images.githubusercontent.com/108006/207713320-a16b5774-3086-4a2e-b95d-43b2412fe162.gif)

**Comment by joreg** [comment]
this looks great!

just one idea: what if there wasn't even a bookmark button, because by default every connection would be saved and get an automatic name: "host:port". being able to change that would still be possible but optional.

then when you say each bookmark needs to have a unique name, that would automatically make sure you don't collect duplicates.

also i think the bookmarks shouldn't be "folded away". i think they are a very good addition that should feature prominently.

**Comment by i-n-g-o** [comment]
Hello.

Nice! Looks great!

I also think that the bookmark-list should not hide away, it is a quite useful feature. It could even be located above the ip:port input fields - but maybe not push the input-fields out of sight completely.

(btw. I am thinking to remove the port-input-field and only have one input for ip:port - but that's a different issue)

Not sure about the auto-saving proposal from joreg, but since it is easy to remove the bookmarks why not. Otherwiese I would propse to add a (+)-icon to store a connection before connecting so the controller-view does not need a bookmak-button.

"Tabs in Root": If you have ParameterGroups at the root-level of your rcp-server and this option is set, then it shows the groups as tabs instead of the ususal foldable group containers.

**Comment by joreg** [comment]
i'm thinking auto-saving, because i cannot think of any situation where i'd not like to save it. it's similar to the default behavior of forms on websites where the next time you fill out your e-mail address it already shows up as a suggestion..

**Comment by i-n-g-o** [comment]
true. seems handy to store every connection.

**Comment by bennigraf** [comment]
Hey there!

Thank you for your remarks. I agree that maybe it just makes sense to always store connections, more like a browser history than a bookmark list. Thus I gave it a try – see below.

 * recent connections are added to the list as you go, sorted most recent to oldest
 * Connections are added only once; if you connect to a service that's already in the list, it'll just be moved to the top, but there will be no duplicate
 * I added a little "favourite" functionality there; when you "star" an item, it will always be sorted on top and can have a name (as I don't know how meaningful the `serverApplicationId` is in the wild)
 * You can manually delete items; however, if you don't delete them, the list will grow indefinetly.

Regarding the number of items: In your experience, with how many servers do you usually interact in real life? For me personally it's only exactly one so far, thus I'm probably not a representative user...

![history](https://user-images.githubusercontent.com/108006/209476361-07706119-4978-4026-aad7-3db4b0acb425.gif)

What do you think of this, good enough for a PR?

Best,
Benjamin.

**Comment by joreg** [comment]
looks perfect to me. regarding number of items i'd opt for three. and then it starts scrolling?Sent from Android deviceOn Dec 25, 2022 18:07, Benjamin Graf ***@***.***> wrote:

Hey there!

Thank you for your remarks. I agree that maybe it just makes sense to always store connections, more like a browser history than a bookmark list. Thus I gave it a try – see below.

recent connections are added to the list as you go, sorted most recent to oldestConnections are added only once; if you connect to a service that's already in the list, it'll just be moved to the top, but there will be no duplicateI added a little "favourite" functionality there; when you "star" an item, it will always be sorted on top and can have a name (as I don't know how meaningful the serverApplicationId is in the wild)You can manually delete items; however, if you don't delete them, the list will grow indefinetly.

Regarding the number of items: In your experience, with how many servers do you usually interact in real life? For me personally it's only exactly one so far, thus I'm probably not a representative user...



What do you think of this, good enough for a PR?

Best,

Benjamin.

**Comment by bennigraf** [comment]
I opened a PR in #16 and posted another screen cast there. I changed some details of the design. When the list grows longer, the dialog will grow up until about 80% of screen height and then allow to scroll it's contents.

**Comment by i-n-g-o** [comment]
Hi.

Cool - thank you so much for this. I added some remarks in the PR.

> Regarding the number of items: In your experience, with how many servers do you usually interact in real life? For me personally it's only exactly one so far, thus I'm probably not a representative user...

For me it is a handful of servers I connect to.

**Comment by i-n-g-o** [comment]
Done.
Please check:
http://client.rabbitcontrol.cc/

**Comment by i-n-g-o** [comment]
Thanks again!

btw. I set the scrollbar to hidden because there were some styling issues on desktop.

**Comment by bennigraf** [comment]
Hey,
Great! You're welcome :-). Cool to see this being merged and I'll look forward to use it once I'm back on location :-). 
Best,
Benjamin.
